docs/how-to-guides/create-a-user-defined-function.md

```diff
@@ -0,0 +1,913 @@
+import io.confluent.ksql.function.udf.UdfParameter;
+
+import java.util.List;
+import java.util.LinkedList;
+import java.util.Iterator;
+
+@UdafDescription(name = "rolling_sum",
+                 author = "example user",
+                 version = "2.0.0",
+                 description = "Maintains a rolling sum of the last 3 integers of a stream.")
+public class RollingSumUdaf {
+
+    private RollingSumUdaf() {
+    }
+
+    @UdafFactory(description = "Sums the previous 3 integers of a stream, discarding the oldest elements as new ones arrive.")
+    public static Udaf<Integer, List<Integer>, Integer> createUdaf() {
+        return new RollingSumUdafImpl();
+    }
+
+    private static class RollingSumUdafImpl implements Udaf<Integer, List<Integer>, Integer> {
+
+        private final int CAPACITY = 3;
+
+        @Override
+        public List<Integer> initialize() {
+            return new LinkedList<Integer>();
+        }
+
+        @Override
+        public List<Integer> aggregate(Integer newValue, List<Integer> aggregateValue) {
+            aggregateValue.add(newValue);
+
+            if (aggregateValue.size() > CAPACITY) {
+                aggregateValue = aggregateValue.subList(1, CAPACITY + 1);
+            }
+
+            return aggregateValue;
+        }
+
+        @Override
+        public Integer map(List<Integer> intermediate) {
+            Iterator<Integer> it = intermediate.iterator();
+            int k = 0;
+
+            while(it.hasNext()) {
+                k += it.next();
+            }
+
+            return k;
+        }
+
+        @Override
+        public List<Integer> merge(List<Integer> aggOne, List<Integer> aggTwo) {
+            return aggOne;
+        }
+    }
+}
+```
+
+There are many things to observe in this class:
+
+- By contrast to scalar and tabular functions, aggregation functions are designated by a static method with the `@UdafFactory` annotation. Because aggregations must implement multiple methods, this helps ksqlDB differentiate aggregations when multiple type signatures are used.
+
+- The static factory method needs to either return `Udaf` or `TableUdaf` (in package `io.confluent.ksql.function.udaf`). The former, which is used in this example, can only be used to aggregate streams into tables, and cannot aggregate tables into other tables. To achieve the latter, use the `TableUdaf`, which derives from `Udaf`, and implement the `undo()` method, too.
+
+- ksqlDB decouples the internal representation of an aggregate from how it is used in an operation. This is very useful because aggregations can maintain complex state, but expose it in a simpler way in a query. In this example, the internal representation is a `LinkedList`, as indicated by the `initialize()` method. But when ksqlDB interacts with the aggregation value, `map()` is called, which sums the values in the list. The `List` is needed to keep a running history of values, but the summed value is needed for the query itself.
```

Comment:
```suggestion
- ksqlDB decouples the internal representation of an aggregate from its use in an operation. This is useful because aggregations can maintain complex state and expose it in a simpler way in a query. In the current example, the internal representation is a `LinkedList`, as indicated by the `initialize()` method. But when ksqlDB interacts with the aggregation value, `map()` is called, which sums the values in the list. The `List` is needed to keep a running history of values, but the summed value is needed for the query itself.
```